raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -2415,17 +2476,30 @@ public void resign(int epoch) {
     @Override
     public Optional<SnapshotWriter<T>> createSnapshot(
         OffsetAndEpoch snapshotId,
-        long lastContainedLogTime
+        long lastContainedLogTimestamp
     ) {
-        return RecordsSnapshotWriter.createWithHeader(
-                () -> log.createNewSnapshot(snapshotId),
-                MAX_BATCH_SIZE_BYTES,
-                memoryPool,
-                time,
-                lastContainedLogTime,
-                CompressionType.NONE,
-                serde
-            );
+        if (!isInitialized()) {
+            throw new IllegalStateException("Cannot create snapshot before the replica has been initialized");
+        }
+
+        return log.createNewSnapshot(snapshotId).map(writer -> {
+            long lastContainedLogOffset = snapshotId.offset() - 1;
```

Comment:
nit: maybe we could have a helper for this logic

Reply:
Yes. I have this issue [KAFKA-14620](https://issues.apache.org/jira/browse/KAFKA-14620) to introduce the `SnapshotId` type. I can fix this on that PR.